tket/src/passes/squash_borrow.rs

```diff
@@ -0,0 +1,708 @@
+        /// The type as which the array was created
+        array_ty: Type,
+        /// The type which the op returns after borrowing from it
+        out_ty: Type,
+    },
+}
+
+/// The ports by which the container array reaches and leaves a
+/// particular borrow or return node.
+#[derive(Clone, Debug)]
+pub struct BorrowFromPorts {
+    /// The port receiving the array before the borrow/return.
+    pub inc: IncomingPort,
+    /// The port returning the array after the borrow/return.
+    pub out: OutgoingPort,
+}
+
+/// Whether a node is a borrow or return, along with the port for
+/// the borrowed value.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum BRAction {
+    /// A borrow action, containing the port on which the borrowed value is output.
+    Borrow(OutgoingPort),
+    /// A return action, containing the port on which the value to return is input.
+    Return(IncomingPort),
+    /// Some other action that requires borrowedness of some indices to be as it
+    /// would be in the unoptimized Hugr. Thus, we cannot optimize any boro-ret or
+    /// ret-boro pair across this op.
+    ///
+    /// The indices are given by the combination of [BorrowReturnPorts::elem_index]
+    /// and any indices in the `Vec` herein.
+    ///
+    /// If the op clobbers even a single index which is definitely unknown (not from
+    /// an [IncomingPort]), then we must assume borrowedness of *all* indices must be
+    /// preserved, so no optimization of any pairs around this op is possible; in such
+    /// case, the outputs of the op can be considered to be fresh arrays so the
+    /// appropriate return value from [IsBorrowReturn::is_borrow_return] is `None`.
+    Clobber(Vec<IncomingPort>),
+}
+
+/// Ports common to a borrow or return op
+#[derive(Debug, Clone)]
+pub struct BorrowReturnPorts {
+    /// Whether this is a borrow or return, and the port for the borrowed value.
+    pub action: BRAction,
+    /// Port on which the index to borrow, return, or clobber, is passed in
+    pub elem_index: IncomingPort,
+    /// The ports by which the container array reaches and leaves this node.
+    pub borrow_from: BorrowFromPorts,
+}
+
+/// An element index to a borrow/return.
+/// If right, i.e. non-constant, then no elision may be possible.
+pub type BorrowIndex = Either<u64, Wire>;
+
+impl IsBorrowReturn for BorrowArray {
+    fn is_borrow_return<H: HugrView>(
+        &self,
+        node: H::Node,
+        hugr: &H,
+    ) -> Result<Option<BorrowReturnPorts>, BorrowAnalysisError> {
+        let op = hugr.get_optype(node);
+
+        let Some(ext_op) = op.as_extension_op() else {
+            return Ok(None);
+        };
+        Ok(match BArrayUnsafeOpDef::from_extension_op(ext_op) {
+            Ok(BArrayUnsafeOpDef::borrow) => {
+                let op = hugr.get_optype(node);
+                let sig = op
+                    .dataflow_signature()
+                    .ok_or_else(|| BorrowAnalysisError::NodeNotDataflow { op: op.clone() })?;
```

Comment:
You've already matched on BArrayUnsafeOpDef::borrow so this should be an `expect` imo